src/NHibernate/Dialect/Function/BitwiseFunctionOperation.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+using NHibernate.Type;
+
+namespace NHibernate.Dialect.Function
+{
+	/// <summary>
+	/// Treats bitwise operations as SQL function calls.
+	/// </summary>
+	[Serializable]
+	public class BitwiseFunctionOperation : ISQLFunction
+	{
+		private readonly string _functionName;
+
+		/// <summary>
+		/// Creates an instance of this class using the provided function name
+		/// </summary>
+		/// <param name="functionName">
+		/// The bitwise function name as defined by the SQL-Dialect
+		/// </param>
+		public BitwiseFunctionOperation(string functionName)
+		{
+			_functionName = functionName;
+		}
+
+		#region ISQLFunction Members
+
+		public IType ReturnType(IType columnType, IMapping mapping)
+		{
+			return NHibernateUtil.Int64;
+		}
+
+		public bool HasArguments
+		{
+			get { return true; }
+		}
+
+		public bool HasParenthesesIfNoArguments
+		{
+			get { return true; }
+		}
+
+		public SqlString Render(IList args, ISessionFactoryImplementor factory)
+		{
+			var filteredArgs = new Queue();
+			foreach (var arg in args)
+			{
+				if (!IsParens(arg.ToString()))
```

Comment:
Thinking about it, it might be just an optimization, for removing parens which would be redundant otherwise, because the function call forcibly adds its own parens.

Reply:
It appears to be a hack for the BitwiseQuery3 test. The args contains four elements instead of two, the first being actually an element, the second an opening parenthesis and the fouth a closing parenthesis. Outputing them all comma-separated could only fail. For this case ignoring the parens yield a good result. Could more elaborated queries yield more complex cases requiring more subtle handling? I do not know for sure. But fiddling with the query, I cannot get such a case.

The native case does not bother with it, because parens seem to appears only as second and fourth argument, while the operator is put only after the first argument. So it get suitably placed, and parens do not creates an issue there.

That seems a bit brittle, but it is needed and it seems there are no known failures with that.